### ISSUE TITLE
IE and EDGE crash

### ISSUE THREAD

**Issue by ExDeaDguY** [ag-grid/ag-grid#2487]
**I'm submitting a ...**  (check one with "x")
```
[x ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/ceolter/ag-grid/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Using Angular 6 & latest version of ag-grid (18.0.1 I believe at the time of posting this)- after upgrading ag-grid from 17.1.1 to 18.x, whenever i navigate to a page that has ag-grid on it, the browser freezes up. I opened dev tools and managed to capture an error before it froze. The error said "onResize doesn't exist" or something of this nature. My TS/JS is not using any resize functions what-so ever. This works fine on chrome but IE and edge no longer work. I reverted back to 17.1.1 and everything was working again. (Aside from the clunky crappy scrolling behavior in IE and EDGE when using a large grid (many columns) with a pinned column(s).

**Expected behavior**
To load and not freeze/crash... I was hoping the update would improve IE and Edge performance, not do the opposite.

**Minimal reproduction of the problem with instructions**
simply add an ag-grid onto an html page within an angular 5/6 application.
Code Example: https://pastebin.com/ZQjU2NpF

**Please tell us about your environment:**
VSCODE (NodeJS, CLI, Etc.), Angular 5/6 (Tried in both).

* **ag-Grid version:** X.X.X
18.0.1, 18.1.0

* **Browser:**
Internet Explorer, MS Edge.

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript, ES6/7 (Used), HTML. etc.

**Comment by ExDeaDguY** [comment]
Has anyone else had this behaviour? Does anyone even read these? Will ag-grid ever work correctly with edge/IE?

**Comment by vitodata** [comment]
I do believe it has to do with the new resizeObserver.js. With Firefox and IE ag-grid starts monitoring the resize using a MutationObserver:

this.mutationsObserver_.observe(document, {
                attributes: **true**,
                childList: true,
                characterData: true,
                subtree: true
            });

The monitoring of the attributes on the document level kills basically the performance, respectivelly freezes the browser. 

We need an urgent fix for this! (just found the reason of the problem today!)

**Comment by macqbat** [comment]
Fully agree with @vitodata. Exactly the same case i have meet (angular 5.2.11, ag-grid 18.0.1/18.1.0, Firefox 61). Chrome works good, IE & Edge didn't check. After commenting out these lines in ./lib/resizeObserver.js:

```javascript
this.mutationsObserver_ = new MutationObserver(this.refresh);
 this.mutationsObserver_.observe(document, {
     attributes: true,
      childList: true,
      characterData: true,
     subtree: true
});
```

in FF everything works fine. 
I'm your strongest supporter in quickly finding a fix :-)

**Comment by dominikg** [comment]
see also https://github.com/ag-grid/ag-grid-angular/issues/221

**Comment by dominikg** [comment]
ResizeObserver is used to dispatch GridSizeChangedEvent, so the first thing i'd do is only register it lazily when someone actually adds a listener for that event instead of eagerly doing it when the grid is created.

Also if there is no native ResizeObserver, using a simple timer loop like ag-grid < 18 did would keep the event feature available without the cost of MutationObserver.

**Comment by dominikg** [comment]
The ResizeObserver implementation in ag-grid seems to have been inspired by a polyfill. In that polyfills repo there is a hint that [MutationObserver in IE11 causes a memory leak](https://github.com/que-etc/resize-observer-polyfill/issues/7#issuecomment-391712517). Another reason not to use it.

workaround: If you don't need GridSizeChanged and ResizeObserver, replace window.ResizeObserver with a dummy implementation. ag-grid picks it up and uses it instead of the MutationObserver based one, still some extra function calls, but no freeze.

There are also other libraries that use extra dom elements and some hackery to detect resizes with scroll events. e.g. https://github.com/wnr/element-resize-detector. One could implement ResizeObserver based on that or adapt the method for ag-grid (maybe less extra elements needed due to the existing structure of ag-grid)

**Comment by vitodata** [comment]
it would be enough not to monitor attribute changes, but I don't know whether attribute changes are of any relevance here:

> this.mutationsObserver_.observe(document, {
>      attributes: **false**,
>       childList: true,
>       characterData: true,
>      subtree: true
> });

**Comment by dominikg** [comment]
they are very relevant (style attribute)

**Comment by picomis** [comment]
safari also freezes

**Comment by harithay** [comment]
If any one looking to solve this issue. On FF, Safari, IE the code goes to ResizeObserverFallback since it does not have ResizeObserver in resizeObserver.js. Adding polyfill ResizeObserver that implements Chrome's ResizeObserver will fix this problem by breaking cycles.

Source: https://developers.google.com/web/updates/2016/10/resizeobserver
Polyfill: https://github.com/pelotoncycle/resize-observer

Tested on IE11, Safari and FF

**Comment by seanlandsman** [comment]
I've just done a beta release for this - ag-grid 18.1.2-beta.1

I'd encourage you all to try and see if this resolves the issue 

thanks!

**Comment by seanlandsman** [comment]
Release 18.1.2 of ag-grid fixes this issue.  Thanks for reporting this.

**Comment by dominikg** [comment]
still freezes in firefox

https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/

console output:
```
08:21:46.812 too much recursion[Weitere Informationen] 2 engineblock-bc7b07e99ec5c6739c766b4898e4cff5acfddc137ccb7218377069c32731f1d0.js%20line%201%20%3E%20eval:1342:68
	DefaultDomRenderer2.prototype.addClass dom_renderer.ts:178:42
	DebugRenderer2.prototype.addClass services.ts:787:4
	NgbDropdownMenu</NgbDropdownMenu.prototype.applyPlacement dropdown.js:55
	NgbDropdownMenu</NgbDropdownMenu.prototype.position dropdown.js:40
	NgbDropdown</NgbDropdown.prototype._positionMenu dropdown.js:206
	NgbDropdown/this._zoneSubscription< dropdown.js:142:72
	EventEmitter.prototype.subscribe/schedulerFn< event_emitter.ts:101:55
	SafeSubscriber.prototype.__tryOrUnsub Subscriber.ts:270:6
	SafeSubscriber.prototype.next Subscriber.ts:212:8
	Subscriber.prototype._next Subscriber.ts:142:4
	Subscriber.prototype.next Subscriber.ts:102:6
	Subject.prototype.next Subject.ts:62:8
	EventEmitter.prototype.emit event_emitter.ts:78:20
	checkStable/< ng_zone.ts:247:39
	ZoneDelegate.prototype.invoke https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:388:17
	Zone.prototype.run https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:138:24
	NgZone.prototype.runOutsideAngular ng_zone.ts:220:12
	checkStable ng_zone.ts:247:10
	onLeave ng_zone.ts:314:2
	onInvokeTask ng_zone.ts:266:8
	ZoneDelegate.prototype.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:420:17
	Zone.prototype.runTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:188:28
	ZoneTask.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:496:24
	invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1540:9
	globalZoneAwareCallback https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1566:17
	positionElements positioning.js:262
	NgbDropdownMenu</NgbDropdownMenu.prototype.position dropdown.js:40
	NgbDropdown</NgbDropdown.prototype._positionMenu dropdown.js:206
	NgbDropdown/this._zoneSubscription< dropdown.js:142:72
	EventEmitter.prototype.subscribe/schedulerFn< event_emitter.ts:101:55
	SafeSubscriber.prototype.__tryOrUnsub Subscriber.ts:270:6
	SafeSubscriber.prototype.next Subscriber.ts:212:8
	Subscriber.prototype._next Subscriber.ts:142:4
	Subscriber.prototype.next Subscriber.ts:102:6
	Subject.prototype.next Subject.ts:62:8
	EventEmitter.prototype.emit event_emitter.ts:78:20
	checkStable/< ng_zone.ts:247:39
	ZoneDelegate.prototype.invoke https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:388:17
	Zone.prototype.run https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:138:24
	NgZone.prototype.runOutsideAngular ng_zone.ts:220:12
	checkStable ng_zone.ts:247:10
	onLeave ng_zone.ts:314:2
	onInvokeTask ng_zone.ts:266:8
	ZoneDelegate.prototype.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:420:17
	Zone.prototype.runTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:188:28
	ZoneTask.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:496:24
	invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1540:9
	globalZoneAwareCallback https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1566:17
	DefaultDomRenderer2.prototype.addClass dom_renderer.ts:178:42
	DebugRenderer2.prototype.addClass services.ts:787:4
	NgbDropdownMenu</NgbDropdownMenu.prototype.applyPlacement dropdown.js:55
	NgbDropdownMenu</NgbDropdownMenu.prototype.position dropdown.js:40
	NgbDropdown</NgbDropdown.prototype._positionMenu dropdown.js:206
	NgbDropdown/this._zoneSubscription< dropdown.js:142:72
	EventEmitter.prototype.subscribe/schedulerFn< event_emitter.ts:101:55
	SafeSubscriber.prototype.__tryOrUnsub Subscriber.ts:270:6
	SafeSubscriber.prototype.next Subscriber.ts:212:8
	Subscriber.prototype._next Subscriber.ts:142:4
	Subscriber.prototype.next Subscriber.ts:102:6
	Subject.prototype.next Subject.ts:62:8
	EventEmitter.prototype.emit event_emitter.ts:78:20
	checkStable/< ng_zone.ts:247:39
	ZoneDelegate.prototype.invoke https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:388:17
	Zone.prototype.run https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:138:24
	NgZone.prototype.runOutsideAngular ng_zone.ts:220:12
	checkStable ng_zone.ts:247:10
	onLeave ng_zone.ts:314:2
	onInvokeTask ng_zone.ts:266:8
	ZoneDelegate.prototype.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:420:17
	Zone.prototype.runTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:188:28
	ZoneTask.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:496:24
	invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1540:9
	globalZoneAwareCallback https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1566:17
	positionElements positioning.js:262
	NgbDropdownMenu</NgbDropdownMenu.prototype.position dropdown.js:40
	NgbDropdown</NgbDropdown.prototype._positionMenu dropdown.js:206
	NgbDropdown/this._zoneSubscription< dropdown.js:142:72
	EventEmitter.prototype.subscribe/schedulerFn< event_emitter.ts:101:55
	SafeSubscriber.prototype.__tryOrUnsub Subscriber.ts:270:6
	SafeSubscriber.prototype.next Subscriber.ts:212:8
	Subscriber.prototype._next Subscriber.ts:142:4
	Subscriber.prototype.next Subscriber.ts:102:6
	Subject.prototype.next Subject.ts:62:8
	EventEmitter.prototype.emit event_emitter.ts:78:20
	checkStable/< ng_zone.ts:247:39
	ZoneDelegate.prototype.invoke https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:388:17
	Zone.prototype.run https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:138:24
	NgZone.prototype.runOutsideAngular ng_zone.ts:220:12
	checkStable ng_zone.ts:247:10
	onLeave ng_zone.ts:314:2
	onInvokeTask ng_zone.ts:266:8
	ZoneDelegate.prototype.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:420:17
	Zone.prototype.runTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:188:28
	ZoneTask.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:496:24
	invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1540:9
	globalZoneAwareCallback https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1566:17
	DefaultDomRenderer2.prototype.removeClass dom_renderer.ts:180:45
	DebugRenderer2.prototype.removeClass services.ts:795:4
	NgbDropdownMenu</NgbDropdownMenu.prototype.applyPlacement dropdown.js:45
	NgbDropdownMenu</NgbDropdownMenu.prototype.position dropdown.js:40
	NgbDropdown</NgbDropdown.prototype._positionMenu dropdown.js:206
	NgbDropdown/this._zoneSubscription< dropdown.js:142:72
	EventEmitter.prototype.subscribe/schedulerFn< event_emitter.ts:101:55
	SafeSubscriber.prototype.__tryOrUnsub Subscriber.ts:270:6
	SafeSubscriber.prototype.next Subscriber.ts:212:8
	Subscriber.prototype._next Subscriber.ts:142:4
	Subscriber.prototype.next Subscriber.ts:102:6
	Subject.prototype.next Subject.ts:62:8
	EventEmitter.prototype.emit event_emitter.ts:78:20
	checkStable/< ng_zone.ts:247:39
	ZoneDelegate.prototype.invoke https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:388:17
	Zone.prototype.run https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:138:24
	NgZone.prototype.runOutsideAngular ng_zone.ts:220:12
	checkStable ng_zone.ts:247:10
	onLeave ng_zone.ts:314:2
	onInvokeTask ng_zone.ts:266:8
	ZoneDelegate.prototype.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:420:17
	Zone.prototype.runTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:188:28
	ZoneTask.invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:496:24
	invokeTask https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1540:9
	globalZoneAwareCallback https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/turbo_modules/zone.js@0.8.26/dist/zone.js:1566:17
	DefaultDomRenderer2.prototype.removeClass dom_renderer.ts:180:45
	DebugRenderer2.prototype.removeClass services.ts:795:4
	NgbDropdownMenu</NgbDropdownMenu.prototype.applyPlacement dropdown.js:45
	NgbDropdownMenu</NgbDropdownMenu.prototype.position dropdown.js:40
	NgbDropdown</NgbDropdown.prototype._positionMenu dropdown.js:206
	NgbDropdown/this._zoneSubscription< dropdown.js:142:72
	EventEmitter.prototype.subscribe/schedulerFn< event_emitter.ts:101:55
```

**Comment by Shemesh** [comment]
still freezes

**Comment by jfpicard1** [comment]
It still an issue in IE 11. We needed to revert to 16.0.1 to make the grid work.

**Comment by ExDeaDguY** [comment]
@jfpicard1 I'm using the following versions according to my package.json:

```
"ag-grid": "^18.1.2",
 "ag-grid-angular": "^18.1.0"
```

These appear to be working great in the following browsers / versions: 
"Microsoft Edge v42.17134.1.0",
"Internet Explorer 11.165..17134.0",
"Chrome v67.0.3396.99 (Official Build) (64-bit)",
"Firefox Quantum v61.0.1 (64-bit)",
and "Opera v54.0.2952.64"

I have not had any app breaking bugs after updating. Keep in mind, you may need to update other packages as well to be sure everything will play nice together.

Site note: Microsoft Edge has scrolling glitches / lag, IE has a bit of lag as well / animation issues. Everything else runs as smooth as butter.

**Comment by dominikg** [comment]
@ExDeaDguY 
opening https://angular-ng-bootstrap-ag-grid-fail-ymr3ut.stackblitz.io/ with firefox 61.01 (macos) and clicking on 'toggle dropdown' causes lots of console outputs with 'too much recursion' like i posted on #issuecomment-408326700 and the dropdown cannot be closed.

what did you do in your app to avoid that?

**Comment by jfpicard1** [comment]
@dominikg 

Your link is freezing in IE11 in my computer. Probably on a infinite loop. The problem is definitively not solved.

Can we reopen that report please ?

**Comment by makinggoodsoftware** [comment]
Hi,



Just to let you know that we are working in a solution for this, we are hoping to have it available in our next release, that should be available in the next 2/3 weeks.



I am afraid that for the time being the only solution would be to downgrade to v17 or to not combine ag-grid with ng-bootstrap



Thanks

**Comment by jfpicard1** [comment]
@makinggoodsoftware 

Just for your information, we use here:
"bootstrap": "^3.3.7",
"ngx-bootstrap": "^3.0.1",

And reverting to v16 solved the problem.

**Comment by DeemaMarwan** [comment]
Hello, 
just wanted to ask if anyone have an update on the IE crashes issue, i am trying to use the new column filtration features (AND/OR) in (18.0.0) ,but cannot upgrade before the crashing issue is solved ,

**Comment by RodSmithAG** [comment]
Hi,

Thank you for your email. We are aware of this issue and have a feature request in our backlog:

AG-1966 - Ag-grid-angular crashes on Firefox/IE/Edge/safari if ng-bootstrap dropdown is clicked 

It is estimated for our next release

You can follow the status of feature requests, bugs and releases using our pipeline: 
https://www.ag-grid.com/ag-grid-pipeline/

If you are an ag-Grid Enterprise Customer, please raise this in our Zendesk support system - contact info@ag-grid.com for access. This is our primary channel for Support.

**Comment by jason2211** [comment]
Also ran into this issue on FF and Edge.  Glad to hear you guys are working on it.  

For anyone else out there, I was able to get the current release 18.1.0 working by commenting out then following lines.  I've haven't noticed any other issues.  I wouldn't recommend doing this in a production app, luckily my app is still in development so this approach was favorable over downgrading while I wait for the next release.

dist\lib\resideObserver.js  (lines 186-189)

```
ResizeObserverController.prototype.connect_ = function () {
        if (this.connected_) {
            return;
        }
        // window.addEventListener("resize", this.refresh);
        // document.addEventListener("DOMSubtreeModified", this.refresh);
        // this.mutationEventsAdded_ = true;
        // this.connected_ = true;
    };
```

**Comment by jdp80** [comment]
Those who are seeing crashes; do you have structural elements that use objects in *ngIf statements?

<div *ngIf="getObject(); let obj">{{obj.foo}}
</div>

In the component,
getObject() { return {foo: 'bar'}; }

For me, IE11 will always hang when getObject() returns any object, if ag-grid is also present. I don’t see a crash if I return string, undefined etc, and it works fine in Chrome.

I get an infinite loop through zone.js globalZoneAwareCallback() where DOMSubtreeModified fires from document.

This is with 18.1.2 enterprise

**Comment by dominikg** [comment]
@jdp80 

your getter returns a new object instance on every call, i think this trips angular change detection...
try sth like this in your component:

```
private _obj = {foo: 'bar'};
get obj() {
  console.log('obj getter called');
  return this._obj;
}
```

and in your template
```
<div *ngIf="obj">{{obj.foo}}</div>
```

**Comment by ExDeaDguY** [comment]
Yeah, so I'm on version 18.1.2 now and ag-grid crashes internet explorer (Freezes it) whenever the <ag-grid-angular> is present. Not sure what changed for me, maybe IE updated or maybe I missed this before? Any new on this yet?

**Comment by jdp80** [comment]
I would go through your templates and check every *ngIf *ngFor etc to see if you’re calling complex functions or, as above, any getter which always returns a new object instead of the same object (possibly with changes). Have a look with Chrome profiler and see if you get hefty cpu usage during the change detection cycle.

Chrome seems to handle complex change detection cycles better than IE, so when ag-grid starts firing multiple change detections it seems to grind IE to a halt. Moving anything complex into cached/template variables run on model changes yielded an immediate improvement (and is probably good practice for Angular in any case)

**Comment by ExDeaDguY** [comment]
Alright thanks, I'll take a look. What I did for now was reverted back to 17.1 again and everything is working fine. The odd thing is, I have 6 or 7 pages, each with an ag-grid-angular on it and each page freezes with IE. Some of the grids are very basic. As soon as I comment out the "<ag-grid-angular />" element, the page runs fine.

By basic, I mean: has a theme class, and is bound to an already filled array. Nothing on the page changes and everything is generally read only after the page does oninit.
